### PR TITLE
Refactor and make way for yarn workspaces

### PR DESCRIPTION
**Pull request by prometheansacrifice**
This PR does not introduce any new functionality and simply makes way for yarn workspaces (and more future setups)

### CODE REVIEW

**Comment by prometheansacrifice** [comment]
Will stop it this for today.

File: src/Esy.re [hunk 1]
```diff
@@ -1,3 +1,4 @@
+/** Esy API **/
```
Comment (wokalski) [inline src/Esy.re]: There are still a couple of promise rejections in this file, is this intentional?

File: src/CheckBucklescriptCompat.re [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    switch (
+      manifestJson |> (field("dependencies", dict(string)) |> optional),
+      manifestJson |> (field("devDependencies", dict(string)) |> optional),
+    ) {
+    | (Some(dependenciesJson), None)
+    | (None, Some(dependenciesJson)) => processDeps(dependenciesJson)
+    | (Some(dependenciesJson), Some(devDependenciesJson)) =>
+      processDeps(mergeDicts(dependenciesJson, devDependenciesJson))
+    | (None, None) =>
+      Error(
```
Comment (wokalski) [inline src/CheckBucklescriptCompat.re]: 👍

File: src/Extension.re [hunk 3]
```diff
@@ -5,27 +5,37 @@ let createClient = (~id, ~name, ~folder) =>
...
     createClient(
       ~id="merlin-language-server",
       ~name="Merlin Language Server",
       ~folder=Workspace.rootPath,
     )
-    |> then_((client: LanguageClient.t) => client.start(.) |> resolve)
+    |> then_(client =>
+         switch (client) {
+         | Ok((client: LanguageClient.t)) => client.start(.) |> resolve
+         | Error(e) => Window.showErrorMessage(e)
```
Comment (wokalski) [inline src/Extension.re]: Nice.

**Comment by prometheansacrifice** [comment]
In the upcoming PRs (probably after your multiworkspace one), I want to turn the Setup pipeline into a separate progress emitter. Motivating: make it easy to test new setups without having to run slow vscode integration tests. And of course, to handle errors better. How about implementing this variants approach to errors there?

**Comment by prometheansacrifice** [comment]
> It'd be great if this could show an internal error dialog (i.e. web-like alert UI, not the notification-like UI) with the details and asked the user to file an issue:

I was thinking about this for setup failures too. As of now I haven't figure a nice way to do this - do you have anything in mind?

**Comment by wokalski** [comment]
I like the idea with a separate progress emitter @prometheansacrifice. No need to do it now.

**Comment by wokalski** [comment]
When it comes to the error I thought about showing a modal for errors instead of the notification like UI.

**Comment by prometheansacrifice** [comment]
> When it comes to the error I thought about showing a modal for errors instead of the notification like UI.

Not sure if there is an API for this. Will dig deeper. Pointers appreciated!

**Comment by wokalski** [comment]
Theres is! https://code.visualstudio.com/api/references/vscode-api

Check showErrorMessage

**Pull request merged**